### PR TITLE
191 scenario#1

### PR DESCRIPTION
**Pull request by perepichai**
issue: https://github.com/ita-social-projects/OoS-Backend/issues/191
implemented part of scenario#1:

-add pop-up when reject apclication (with required textarea):
image

-display reject reason msg:
image

### CODE REVIEW

File: src/app/shared/components/reject-modal-window/reject-modal-window.component.html [hunk 1]
```diff
@@ -0,0 +1,27 @@
+<div fxLayout="column" fxLayoutAlign="center center">
```
Comment (dmitryy90) [inline src/app/shared/components/reject-modal-window/reject-modal-window.component.html:1]: indentations

File: src/app/shared/components/reject-modal-window/reject-modal-window.component.ts [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+  constructor(
+    private formBuilder: FormBuilder,
+    private store: Store,
+    public dialogRef: MatDialogRef<RejectModalWindowComponent>,
+    @Inject(MAT_DIALOG_DATA) public data: DialogData) {
+    this.ReasonFormGroup = this.formBuilder.group({
+      description: new FormControl('', Validators.required)
+    });
+   }
```
Comment (dmitryy90) [inline src/app/shared/components/reject-modal-window/reject-modal-window.component.ts:38]: indentations

File: src/app/shared/models/application.model.ts [hunk 3]
```diff
@@ -22,9 +23,11 @@ export class Application {
 export class ApplicationUpdate {
   id: string;
   status: string;
+  rejectionMessage: string
 
-  constructor(id, status) {
+  constructor(id, status, rejectionMessage?) {
```
Comment (dmitryy90) [inline src/app/shared/models/application.model.ts:28]: types?

File: src/app/shell/personal-cabinet/applications/application-card/application-card.component.ts [hunk 4]
```diff
@@ -58,8 +68,27 @@ export class ApplicationCardComponent implements OnInit {
    * This method emit on deny action
    * @param Application application
    */
-  onReject(application: Application): void {
-    this.rejected.emit(application);
+  // onReject(application: Application): void {
```
Comment (dmitryy90) [inline src/app/shell/personal-cabinet/applications/application-card/application-card.component.ts:71]: do we still need this?

**Review by dmitryy90 — CHANGES_REQUESTED** [review]

File: src/app/shared/components/reject-modal-window/reject-modal-window.component.html [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+  <h2 mat-dialog-title>{{ modalTitle }}</h2>
+  <p class="confirmation"> {{ modalDescription }} 
+  </p>
+
+  <form [formGroup]="ReasonFormGroup" fxLayout='column' class="reject-form">
```
Comment (litvinets) [inline src/app/shared/components/reject-modal-window/reject-modal-window.component.html:6]: use please double quotes in the template

**Comment by litvinets** [comment]
I see that you made the rejected message required, then obviously the rejected status will be always supported with teh rejected message.
Then:
1) You should display message only for rejected status
2) You should remove "Не вказано" from the template
image

**Comment by litvinets** [comment]
The validation styles is incorrect, the textarea should have red border 
image

File: src/app/shared/components/reject-modal-window/reject-modal-window.component.ts (hunk 2 above)
Comment (perepichai) [reply src/app/shared/components/reject-modal-window/reject-modal-window.component.ts]: is updated

File: src/app/shared/components/reject-modal-window/reject-modal-window.component.ts [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+
+  ngOnInit(): void {
+    this.modalTitle = 'ВІДМОВИТИ',
```
[the remark this answers is not in the document]
Comment (perepichai) [reply src/app/shared/components/reject-modal-window/reject-modal-window.component.ts]: is updated as well

File: src/app/shared/models/application.model.ts (hunk 3 above)
Comment (perepichai) [reply src/app/shared/models/application.model.ts]: the types is added

File: src/app/shell/personal-cabinet/applications/application-card/application-card.component.ts (hunk 4 above)
Comment (perepichai) [reply src/app/shell/personal-cabinet/applications/application-card/application-card.component.ts]: removed this part

File: src/app/shared/components/reject-modal-window/reject-modal-window.component.html (hunk 5 above)
Comment (perepichai) [reply src/app/shared/components/reject-modal-window/reject-modal-window.component.html]: is updated, thanks

**Comment by perepichai** [comment]
>

**Comment by perepichai** [comment]
improve validations styles @litvinets

**Review by perepichai — COMMENTED** [review]
was following your comments

File: src/app/shared/components/reject-modal-window/reject-modal-window.component.html [hunk 7]
```diff
@@ -0,0 +1,27 @@
+<div fxLayout="column" fxLayoutAlign="center center">
+  <h2 mat-dialog-title>{{ modalTitle }}</h2>
+  <p class="confirmation"> {{ modalDescription }} 
+  </p>
+
+  <form [formGroup]="ReasonFormGroup" fxLayout="column" class="reject-form">
+  
+    <label class="step-label"> Вкажіть причину відмови <span class="step-required">*</span></label>
+    <mat-form-field appPlaceholderStyling floatLabel="never" appearance="none">
+      <div [ngClass]="(ReasonFormGroup.get('description').invalid && ReasonFormGroup.get('description').touched) ? 'redBorder' : '' " class="wrapper-border">
```
Comment (litvinets) [inline src/app/shared/components/reject-modal-window/reject-modal-window.component.html:10]: First of all, when you do firstly ngClass, then later you add just class, then the first class will be not applied.
So, in this case, this div will have ONLY wrapper-border, but "red-border" will never apply. 
The thing you want to do, is 
```suggestion
      <div [ngClass]="(ReasonFormGroup.get('description').invalid && ReasonFormGroup.get('description').touched) ? 'redBorder wrapper-border' : '' wrapper-border">
```
Secondly, I truly believe that red borders are already handled by validation.scss. You need to apply it not for the div, but for the textarea. I think you should reconsider the templated, get rid of this div-wrapper and apply classes directly to textarea.  Take a look, how textarea are implemented in other components.

File: src/app/shared/components/reject-modal-window/reject-modal-window.component.html [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+
+    <app-validation-hint-for-input type="requiredField"
+      [invalid]="ReasonFormGroup.get('description').invalid && ReasonFormGroup.get('description').touched">
+    </app-validation-hint-for-input>
+  </form>
+
+  <mat-dialog-actions fxLayout="row wrap" fxLayoutAlign="center center" class="footer">
+    <button class="btn" mat-button
+      [mat-dialog-close]="ReasonFormGroup.get('description').value"
+      cdkFocusInitial [disabled]="ReasonFormGroup.get('description').value === ''">ПІДТВЕРДИТИ</button>
```
Comment (litvinets) [inline src/app/shared/components/reject-modal-window/reject-modal-window.component.html:24]:
```suggestion
      cdkFocusInitial [disabled]="!ReasonFormGroup.get('description').value">ПІДТВЕРДИТИ</button>
```

File: src/app/shared/components/reject-modal-window/reject-modal-window.component.spec.ts [hunk 9]
```diff
@@ -0,0 +1,71 @@
+import { ComponentFixture, TestBed } from '@angular/core/testing';
+
```
Comment (litvinets) [inline src/app/shared/components/reject-modal-window/reject-modal-window.component.spec.ts:2]: remove please this empty line

**Review by litvinets — CHANGES_REQUESTED** [review]

File: src/app/shared/components/reject-modal-window/reject-modal-window.component.html (hunk 7 above)
Comment (perepichai) [reply src/app/shared/components/reject-modal-window/reject-modal-window.component.html:10]: Thank you for the explanation. regarding wrapper div, before implement current aproach I just followed other components example, they have wrapper div. However, I realized that the border doesn't appear when textarea is empty. So I checked the validation form scss file and got 'redBoder' class. I thought it exist for cases like my)
Now I've removed wrapper div and added one more class for the textarea. So it works perfectly now) Please confirm if we can leave this one without wrapper div. Thanks again

File: src/app/shared/components/reject-modal-window/reject-modal-window.component.html (hunk 8 above)
Comment (perepichai) [reply src/app/shared/components/reject-modal-window/reject-modal-window.component.html:24]: cool, thanks.

File: src/app/shared/components/reject-modal-window/reject-modal-window.component.spec.ts (hunk 9 above)
Comment (perepichai) [reply src/app/shared/components/reject-modal-window/reject-modal-window.component.spec.ts]: done it

File: src/app/shared/components/reject-modal-window/reject-modal-window.component.ts [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+  selector: 'app-reject-modal-window',
+  templateUrl: './reject-modal-window.component.html',
+  styleUrls: ['./reject-modal-window.component.scss'],
+})
+export class RejectModalWindowComponent implements OnInit {
+  readonly constants: typeof Constants = Constants;
+
+  modalTitle = 'ВІДМОВИТИ';
+  modalDescription =
+    'Ви впевнені, що хочете перевести заяву у статус ”Відмовлено”?';
```
Comment (litvinets) [inline src/app/shared/components/reject-modal-window/reject-modal-window.component.ts:22]: align it on the same line please

File: src/app/shared/components/reject-modal-window/reject-modal-window.component.ts [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+
+  ReasonFormGroup: FormGroup;
+
+  @Input() application: Application;
```
Comment (litvinets) [inline src/app/shared/components/reject-modal-window/reject-modal-window.component.ts:26]: usually, @Input and @Output go first, after this is a declaration of properties. Check the code style angular guideline. We are not very strict with that, but Input/Output should go top

File: src/app/shell/personal-cabinet/applications/application-card/application-card.component.ts [hunk 12]
```diff
@@ -23,6 +29,7 @@ export class ApplicationCardComponent implements OnInit {
   childAge: string;
   deviceToogle: boolean;
   infoShowToggle: boolean = false;
+  ReasonFormGroup: FormGroup;
```
Comment (litvinets) [inline src/app/shell/personal-cabinet/applications/application-card/application-card.component.ts:32]: do you need this?

File: src/app/shell/personal-cabinet/applications/application-card/application-card.component.ts [hunk 13]
```diff
@@ -6,12 +6,18 @@ import { ApplicationTitles, ApplicationStatusDescription } from 'src/app/shared/
...
+import { FormControl, FormGroup } from '@angular/forms';
+
+
 @Component({
   selector: 'app-application-card',
   templateUrl: './application-card.component.html',
   styleUrls: ['./application-card.component.scss']
 })
 
+
```
Comment (litvinets) [inline src/app/shell/personal-cabinet/applications/application-card/application-card.component.ts:20]: remove this extra line please

File: src/app/shell/personal-cabinet/applications/application-card/application-card.component.ts [hunk 14]
```diff
@@ -55,11 +65,22 @@ export class ApplicationCardComponent implements OnInit {
...
-   * This method emit on deny action
+   * This method emit reject Application
    * @param Application application
    */
-  onReject(application: Application): void {
-    this.rejected.emit(application);
+    onReject(application: Application): void {
+    const dialogRef = this.matDialog.open(RejectModalWindowComponent, {
+      width: '330px',
+      data: {
```
Comment (litvinets) [inline src/app/shell/personal-cabinet/applications/application-card/application-card.component.ts:74]: I belive you just can remove it if you're not passing any data there?

**Review by litvinets — CHANGES_REQUESTED** [review]

File: src/app/shared/components/reject-modal-window/reject-modal-window.component.ts [hunk 15]
```diff
@@ -0,0 +1,42 @@
...
+  styleUrls: ['./reject-modal-window.component.scss'],
+})
+export class RejectModalWindowComponent implements OnInit {
+  readonly constants: typeof Constants = Constants;
+
+  modalTitle = 'ВІДМОВИТИ';
+  modalDescription =
+    'Ви впевнені, що хочете перевести заяву у статус ”Відмовлено”?';
+
+  ReasonFormGroup: FormGroup;
```
Comment (litvinets) [inline src/app/shared/components/reject-modal-window/reject-modal-window.component.ts:24]: Well, in your case there is one reactive field - description, that's all, so I believe you don't need the formgroup? FormGroup you need usually when you wanna unite several reactive fields in one form, for example, name, surname, date of birth. Then, you can have formgourp where you can track the changing of all the fields. But in your case, there is only one, so probably get rid of FormGroup and use one FormControl.  

```suggestion
  ReasonFormControl= new FormControl('');
```

and then in template, you can remove everything related to the form group and add [formControl]="ReasonFormControl" to the textarea.

**Review by litvinets — CHANGES_REQUESTED** [review]

File: src/app/shared/components/reject-modal-window/reject-modal-window.component.html (hunk 7 above)
Comment (litvinets) [reply src/app/shared/components/reject-modal-window/reject-modal-window.component.html]: I think now it looks good! If it works correct, then good job:)

File: src/app/shell/personal-cabinet/applications/application-card/application-card.component.ts (hunk 14 above)
Comment (perepichai) [reply src/app/shell/personal-cabinet/applications/application-card/application-card.component.ts]: removed it

File: src/app/shared/components/reject-modal-window/reject-modal-window.component.ts (hunk 10 above)
Comment (perepichai) [reply src/app/shared/components/reject-modal-window/reject-modal-window.component.ts]: fixed it

File: src/app/shared/components/reject-modal-window/reject-modal-window.component.ts (hunk 11 above)
Comment (perepichai) [reply src/app/shared/components/reject-modal-window/reject-modal-window.component.ts]: updated this

File: src/app/shell/personal-cabinet/applications/application-card/application-card.component.ts (hunk 12 above)
Comment (perepichai) [reply src/app/shell/personal-cabinet/applications/application-card/application-card.component.ts]: removed it

File: src/app/shell/personal-cabinet/applications/application-card/application-card.component.ts (hunk 13 above)
Comment (perepichai) [reply src/app/shell/personal-cabinet/applications/application-card/application-card.component.ts]: done it

File: src/app/shared/components/reject-modal-window/reject-modal-window.component.ts (hunk 15 above)
Comment (perepichai) [reply src/app/shared/components/reject-modal-window/reject-modal-window.component.ts]: thanks, it is updated

**Review by litvinets — APPROVED** [review]

**Review by dmitryy90 — APPROVED** [review]

**Pull request merged**